internal/abi/fastly/hostcalls_guest.go

```diff
@@ -636,13 +638,19 @@ func fastlyHTTPReqDownstreamTLSClientHello(
 // DownstreamTLSClientHello returns the ClientHello message sent by the client
 // in the singleton downstream request, if any.
 func DownstreamTLSClientHello() ([]byte, error) {
-	buf := prim.NewWriteBuffer(defaultBufferLen)
-
-	if err := fastlyHTTPReqDownstreamTLSClientHello(
+	n := DefaultLargeBufLen
+alloc:
+	buf := prim.NewWriteBuffer(n) // Longest (~132,000); typically < 2^14; RFC https://datatracker.ietf.org/doc/html/rfc8446#section-4.1.2
+	status := fastlyHTTPReqDownstreamTLSClientHello(
 		prim.ToPointer(buf.Char8Pointer()),
 		buf.Cap(),
 		prim.ToPointer(buf.NPointer()),
-	).toError(); err != nil {
+	)
+	if status == FastlyStatusBufLen && buf.NValue() > 0 {
+		n = int(buf.NValue())
+		goto alloc // avoid all the allocations of a new stack, etc.
+	}
```

Comment:
Yeah, I was just thinking of how we do it in other places like https://github.com/fastly/compute-sdk-go/blob/bcb1f0a4f338fcfc56088b3986308f7bc6498c84/internal/abi/fastly/hostcalls_guest.go#L2798-L2815

It reassigns (but doesn't re-declare) `buf` and `status` by calling the functions again without a `goto`.

I might be the one misreading (or misunderstanding)... the comment suggested to me that the `goto` is necessary to avoid allocations.

Reply:
I think the comment just lacks context in that if we did a function closure to call the hostcall again that it would probably undo a good amount of our memory savings.

Reply:
I didn't like repeating the call in a nested conditional

Reply:
Ah, ok!  Understood now, thanks.

Reply:
I've updated these comments to try to clarify.